RC_Channel: add RC_OPTIONS bit to clear RC override by RC input change

tatsuy opened the pull request:
This PR introduces a new option to clear the RC override, allowing the pilot to regain control by operating roll, pitch, throttle, or yaw during an RC override situation. While this functionality was previously possible using RCx_OPTION(46), the new option provides a more direct and safer method for pilots to cancel overrides using basic RC inputs.

## Required parameters
The following settings are required
`RC_OPTIONS = 16384` (bit 14, `CLEAR_OVERRIDES_BY_RC`) enables the feature
`RCx_OPTION = 46` (`RC_OVERRIDE_ENABLE`) Re-enable RC overrides

## Per-vehicle behavior
Vehicles override `RC_Channels::has_pilot_input_for_override_clear()` to define which axes count as pilot input:

| Vehicle | Axes that clear an active override |
|---|---|
| Copter / Plane / Blimp | roll, pitch, yaw out of trim deadzone, or throttle moved beyond deadzone since the override began |
| Rover | steering (rcmap roll), throttle; plus lateral on omni rovers |
| Sub | forward, lateral, yaw, plus throttle (= vertical) |

## Testing
Tested manually in SITL with Copter, Rover, and Plane.
Added autotest:
`RCOverridesClearByPilotInput` on Copter (all four primary axes each clear) and Rover (steering/throttle clear; pitch/yaw must not), plus held-while-neutral and bit-off regression checks.

muramura commented:
I don't think it is possible for the FC to take control of the CC without permission.
If the CC is abnormal and you cannot get a heartbeat from the CC, you can take control of the CC by judging it abnormal.
If the CC is the primary, the FC is the limbs.
If the CC is connected, it is better to keep the FC as a limb.

tatsuy commented:
@muramura
In this PR, the AP's permission to clear Override can be set in RC_OPTIONS.

rmackay9 commented:
Txs for this.  In general I think it is good practice for the pilot holding an RC to be able to take control away from a companion computer easily.  It makes sense to make it easy for a pilot to do that by simply moving the sticks.

rmackay9 commented on libraries/RC_Channel/RC_Channel.cpp:
```diff
@@ -1313,3 +1313,20 @@ void RC_Channels::convert_options(const RC_Channel::aux_func_t old_option, const
+
+bool RC_Channel::radio_input_changed()
+{
+    if (!rc().ignore_receiver()) {
```
I wonder if we should add, "rc().has_had_rc_receiver() &&" to this line.

rmackay9 replied:
An alternative implementation that would allow avoiding two calls to hal.rcin->read() would be to add a "bool rc_in_changed" member and then update it from within RC_Channel::update().  Something like below although it would be good to get opinions from other developers on whether this is useful or not.

```
// read input from hal.rcin or overrides
bool RC_Channel::update(void)
{
    const rc_in = hal.rcin->read(ch_in);
    rc_in_changed = radio_input_changed(rc_in);

    if (has_override() && !rc().ignore_overrides()) {
        radio_in = override_value;
    } else if (rc().has_had_rc_receiver() && !rc().ignore_receiver()) {
        radio_in = rc_in;
    } else {
        return false;
    }

    if (type_in == RC_CHANNEL_TYPE_RANGE) {
        control_in = pwm_to_range();
    } else {
        //RC_CHANNEL_TYPE_ANGLE
        control_in = pwm_to_angle();
    }

    return true;
}
```

tatsuy replied:
yes, two calls to hal.rcin->read() are not good.
I accept your suggestion.

rmackay9 commented on libraries/RC_Channel/RC_Channel.h:
```diff
@@ -330,6 +333,8 @@ class RC_Channel {
 
     // pwm is stored here
     int16_t     radio_in;
+    // previous radio_in, used to ignore small changes
```
we should specify that "previous_radio_in" is only used by radio_input_changed().

tatsuy replied:
renamed

rmackay9 commented on libraries/RC_Channel/RC_Channel.h:
```diff
@@ -511,6 +516,22 @@ class RC_Channels {
         return get_singleton() != nullptr && (_options & uint32_t(Option::USE_CRSF_LQ_AS_RSSI)) != 0;
     }
 
+    bool clear_overrides_by_roll(void) const {
```
I think these accessors and clear_overrides_by_rc can be made protected or private because they are only ever used from within RC_Channel**s**.

Do we need to have bits and accessors for roll, pitch, yaw and throttle or could we just have a single bit and accessor to cover them all?  It seems sufficient to just allow pilots to retake control if they change any of the 4.

tatsuy replied:
Separating them allows for more flexible development, but also increases risk.
My requirements can be met with a single bit.
I made it private and change to a single bit.

rmackay9 commented on libraries/RC_Channel/RC_Channels.cpp:
```diff
@@ -78,6 +79,11 @@ bool RC_Channels::read_input(void)
     last_update_ms = AP_HAL::millis();
 
+    // clear RC override by the pilot input if RC_OPTION is configured
+    if (has_active_overrides()) {
```
you could move this "if (has_active_overrides()) {" into clear_overrides_by_rc() although it's not a blocker.

tatsuy replied:
fixed

rmackay9 commented on libraries/RC_Channel/RC_Channels.cpp:
```diff
@@ -231,6 +237,21 @@ bool RC_Channels::flight_mode_channel_conflicts_with_rc_option() const
+        (rc().clear_overrides_by_throttle() && channel(AP::rcmap()->throttle() - 1)->radio_input_changed()) ||
+        (rc().clear_overrides_by_yaw() && channel(AP::rcmap()->yaw() - 1)->radio_input_changed())) {
+        set_gcs_overrides_enabled(false);
+        gcs().send_text(MAV_SEVERITY_NOTICE, "Clear overrides by pilot input");
```
This comment should probably include "RC" somewhere.  So perhaps "RC overrides cleared by pilot input"

tatsuy replied:
fixed

tatsuy commented:
@rmackay9 Thank you for your reviews.
I've added a commit to make the changes easier to understand, but I'm going to squash it eventually.

rmackay9 commented on libraries/RC_Channel/RC_Channel.cpp:
```diff
@@ -1313,3 +1316,19 @@ void RC_Channels::convert_options(const RC_Channel::aux_func_t old_option, const
+
+bool RC_Channel::radio_input_changed(uint16_t rc_in)
+{
+    if (rc().has_had_rc_receiver() && !rc().ignore_receiver()) {
```
Going against my previous advice, I'm not sure if we should have the, "rc().has_had_rc_receiver()" part.

tatsuy replied:
OK
radio_input_changed function is called when only has_had_rc_receiver() is true because it is checked at the beginning of read_input.
so I removed it again.

rmackay9 commented on libraries/RC_Channel/RC_Channels_VarInfo.h:
```diff
@@ -89,7 +89,7 @@ const AP_Param::GroupInfo RC_Channels::var_info[] = {
     // @Description: RC input options
     // @User: Advanced
-    // @Bitmask: 0:Ignore RC Receiver, 1:Ignore MAVLink Overrides, 2:Ignore Receiver Failsafe bit but allow other RC failsafes if setup, 3:FPort Pad, 4:Log RC input bytes, 5:Arming check throttle for 0 input, 6:Skip the arming check for neutral Roll/Pitch/Yaw sticks, 7:Allow Switch reverse, 8:Use passthrough for CRSF telemetry, 9:Suppress CRSF mode/rate message for ELRS systems,10:Enable multiple ...
+    // @Bitmask: 0:Ignore RC Receiver, 1:Ignore MAVLink Overrides, 2:Ignore Receiver Failsafe bit but allow other RC failsafes if setup, 3:FPort Pad, 4:Log RC input bytes, 5:Arming check throttle for 0 input, 6:Skip the arming check for neutral Roll/Pitch/Yaw sticks, 7:Allow Switch reverse, 8:Use passthrough for CRSF telemetry, 9:Suppress CRSF mode/rate message for ELRS systems, 10:Enable multipl ...
```
probably need to change the description now that the option bits have been combined.

tatsuy replied:
fixed

rmackay9 commented on libraries/RC_Channel/RC_Channel.h:
```diff
@@ -601,6 +608,13 @@ class RC_Channels {
+    // clear overrides by radio input
+    void clear_overrides_by_rc_in();
+
+    bool clear_overrides_by_rc(void) const {
```
my guess is that other devs may say we don't need this small function that is only called from one place.  Also its name is so close to the method above that it is easy to get confused about which is which.  I'd say we should remove this one but I don't feel strongly about it.

tatsuy replied:
I agree with you.
In fact, I had trouble naming the function.

tatsuy commented:
@rmackay9 
I have used this with real rover.
It works well and it's useful.
Can you check this again?

peterbarker commented:
This  sounds like a good idea to me.

@tatsuy did you want to rebase this?

tatsuy commented:
@peterbarker Thank you for your feedback. I have rebased.

peterbarker commented on libraries/RC_Channel/RC_Channel.cpp:
```diff
@@ -30,7 +30,6 @@ extern const AP_HAL::HAL& hal;
 #include <AP_Math/AP_Math.h>
 
 #include "RC_Channel.h"
-#include <GCS_MAVLink/GCS.h>
```
Still need this include; `GCS_SEND_TEXT` is used

tatsuy replied:
ok, keep it there.

peterbarker commented on libraries/RC_Channel/RC_Channel.h:
```diff
@@ -376,9 +377,12 @@ class RC_Channel {
 
     // value generated from PWM normalised to configured scale
-    int16_t    control_in;
+    int16_t     control_in;
```
bad whitespace change

tatsuy replied:
I got it

peterbarker commented on libraries/RC_Channel/RC_Channel.h:
```diff
@@ -651,6 +659,10 @@ class RC_Channels {
     void set_aux_cached(RC_Channel::AUX_FUNC aux_fn, RC_Channel::AuxSwitchPos pos);
 #endif
+
+    bool clear_overrides_by_rc(void) const {
```
Replace this with a call to `option_is_enabled`

tatsuy replied:
Replaced it with a call to option_is_enabled. Thanks for pointing it out.

peterbarker commented on libraries/RC_Channel/RC_Channels.cpp:
```diff
@@ -89,6 +90,20 @@ bool RC_Channels::read_input(void)
+        // clear RC overrides
+        set_gcs_overrides_enabled(false);
+        gcs().send_text(MAV_SEVERITY_NOTICE, "RC overrides cleared by pilot input");
+        if (!rc().find_channel_for_option(RC_Channel::AUX_FUNC::RC_OVERRIDE_ENABLE)) {
+            gcs().send_text(MAV_SEVERITY_INFO, "Set RCx_OPTION=%d to re-enable RC overrides", (uint16_t)RC_Channel::AUX_FUNC::RC_OVERRIDE_ENABLE);
```
```suggestion
```

aux functions can come from many places now - including from the Aux Functions tab in MissionPlanner.

tatsuy replied:
fixed it

peterbarker commented on libraries/RC_Channel/RC_Channels.cpp:
```diff
@@ -89,6 +90,20 @@ bool RC_Channels::read_input(void)
         success |= channel(i)->update();
     }
 
+    // check if clear overrides by RC is allowed by RC_OPTIONS and any change in RC input during RC overrides
```
create a subroutine for this

tatsuy replied:
I've added the `should_clear_overrides` subroutine as suggested.

peterbarker commented on libraries/RC_Channel/RC_Channels.cpp:
```diff
@@ -89,6 +90,20 @@ bool RC_Channels::read_input(void)
 
+    // check if clear overrides by RC is allowed by RC_OPTIONS and any change in RC input during RC overrides
+    if (rc().clear_overrides_by_rc() && has_active_overrides() &&
+        (channel(AP::rcmap()->roll() - 1)->rc_in_changed ||
```
Why don't we just check to see if any of them are outside their deadzones - with the exception of throttle, which we eithre don't check or *do* check for a change...

... and maybe just have a `uint16_t last_throttle` rather than the `rc_in_changed` infrastructure.

We should probably also very that RC input is actually valid before checking any channel value!

tatsuy commented:
@peterbarker Thank you for the review.
I have made the necessary changes based on your feedback.
I also confirmed the changes through SITL testing. Could you please review the updates again?

peterbarker reviewed (commented):
This is looking good, and a good feature IMO.

tatsuy commented on libraries/RC_Channel/RC_Channel.h:
```diff
@@ -378,6 +380,7 @@ class RC_Channel {
 
     // pwm is stored here
     int16_t     radio_in;
+    int16_t     raw_radio_in;
```
added the comment

tatsuy commented on libraries/RC_Channel/RC_Channels.cpp:
```diff
@@ -90,9 +91,37 @@ bool RC_Channels::read_input(void)
     }
 
+    // check if clear overrides by RC is allowed by RC_OPTIONS and any change in RC input during RC overrides
+    if (should_clear_overrides()) {
```
yes, it doesn’t directly clear override values, so I've renamed it to should_ignore_overrides.

peterbarker commented:
We tested this today on a quadcopter and it worked very well - exactly what we were expecting.

peterbarker reviewed (approved):
LGTM

tridge commented on libraries/RC_Channel/RC_Channels.cpp:
```diff
@@ -90,9 +92,36 @@ bool RC_Channels::read_input(void)
+    // check for pilot input exceeding the deadzone on roll, pitch, or yaw
+    if (!get_roll_channel().in_raw_trim_dz() ||
+        !get_pitch_channel().in_raw_trim_dz() ||
+        !get_yaw_channel().in_raw_trim_dz()) {
```
should test on rover, needs to work with only roll and throttle

tatsuy replied:
thanks! I fixed it. pitch/yaw are intentionally ignored.

LupusTheCanine commented on libraries/RC_Channel/RC_Channels.cpp:
```diff
@@ -90,9 +92,36 @@ bool RC_Channels::read_input(void)
+    }
+
+    // check for pilot input exceeding the deadzone on roll, pitch, or yaw
+    if (!get_roll_channel().in_raw_trim_dz() ||
+        !get_pitch_channel().in_raw_trim_dz() ||
+        !get_yaw_channel().in_raw_trim_dz()) {
+        return true;
+    }
```
How about a bitmask of RC channels that trigger takeover?

tatsuy replied:
Thank you for the review. I added a virtual RC_Channels::has_pilot_input_for_override_clear().

tatsuy replied:
@LupusTheCanine 
Thanks for the comment.
A bitmask would provide more flexibility, but this PR intentionally keeps the user-facing option simple: one RC_OPTIONS bit enables/disables the feature.
https://github.com/ArduPilot/ardupilot/pull/19844#discussion_r798198744

peterbarker commented:
Ping @tatsuy this was a good one, just need to fix it to understand different vehicles want different channels to disable the overrdes.

tatsuy committed:
Rover: clear RC overrides on steering/throttle/lateral input

Override has_pilot_input_for_override_clear() to use steering (rcmap
roll) and throttle, plus lateral on omni rovers, instead of the
Copter-style roll/pitch/yaw/throttle set.

tatsuy committed:
ArduSub: clear RC overrides on forward/lateral/yaw/vertical input

Override has_pilot_input_for_override_clear() to use Sub's joystick
axes (forward, lateral, yaw, plus throttle mapped to vertical thrust)
instead of the Copter-style roll/pitch/yaw/throttle set.

tatsuy committed:
RC_Channel: add RC_OPTIONS bit to clear RC override by RC input change

Set CLEAR_OVERRIDES_BY_RC (bit 14) to clear an active MAVLink override
when the pilot moves the RC sticks beyond their deadzones.  Vehicles
override has_pilot_input_for_override_clear() to define which axes
count; default matches Copter/Plane/Blimp.

tatsuy committed:
autotest: add tests for RC_OPTIONS bit 14 (CLEAR_OVERRIDES_BY_RC)

Cover Copter (roll/pitch/yaw/throttle each clear) and Rover
(steering/throttle clear; pitch/yaw do not), plus held-while-neutral
and bit-off regression checks.

tatsuy commented:
@peterbarker Thanks for the feedback, sorry for the long delay.
Rebased and reworked.

- Added a virtual `RC_Channels::has_pilot_input_for_override_clear()`.
Overrides are now cleared based on the following vehicle-specific pilot inputs:
Copter/Plane/Blimp: roll/pitch/yaw/throttle input
Rover: steering and throttle, plus lateral input for omni rovers. Pitch and yaw are intentionally ignored.
ArduSub: forward/lateral/yaw/vertical input

- Added autotest for Copte and Rover
Copter: verifies that each of the four primary axes clears active overrides
Rover: verifies that steering and throttle clear active overrides, while pitch and yaw do not

Tested manually in SITL with Copter, Rover, and Plane.

IamPete1 commented on libraries/RC_Channel/RC_Channels.cpp:
```diff
@@ -106,11 +108,51 @@ bool RC_Channels::read_input(void)
+bool RC_Channels::throttle_moved_since_override_start() const
+{
+    const RC_Channel &thr = get_throttle_channel();
+    return abs(thr.get_raw_radio_in() - override_start_throttle) > thr.get_dead_zone();
```
Should check that `override_start_throttle` is not -1. Raw radio might be 0 so -1 would be within the deadzone.

tatsuy replied:
Thanks, good point.
I added an explicit guard:

if (override_start_throttle < 0) {
    return false;
}

I think override_start_throttle == -1 should not reach this comparison because should_ignore_overrides() first checks has_active_overrides(), and the first active override stores the throttle PWM.
Still, I agree it would be better to add this check.

peterbarker replied:
Did you add the check?

tatsuy replied:
Yes, I added the check here.
https://github.com/ArduPilot/ardupilot/pull/19844/changes#diff-9fe83d8a1f3ccb4c9107664d6c6ae867193a141f4f76221d6bcc1248d2279334R137-R139

IamPete1 reviewed (commented):
How does the pilot re-enable overrides again after taking over?

I'm a little uneasy that it is just the main input channels. There can be some important stuff on switches. The pilot might instinctively try and change modes for example, which would not clear the overrides. But the param description does make it clear that it is "stick" inputs, so maybe its OK.

tatsuy commented:
@IamPete1 
> How does the pilot re-enable overrides again after taking over?

The pilot can use `RCx_OPTION=46` (RC Override Enable) to re-enable overrides again.

peterbarker commented:
We'll test this at SV

rmackay9 commented:
At today's dev call @peterbarker said he was keen to test it this weekend so I've added it to the DevCall again for next week

rmackay9 commented:
oh darn, merging PR https://github.com/ArduPilot/ardupilot/pull/33275 has caused merged conflicts, sorry!

rmackay9 reviewed (approved):
Hi @tatsuy sorry for the delay on this.  Could you resolve the merge conflicts and then we will merge immediately!

tatsuy commented:
@rmackay9 OK, I resolved the merge conflicts.

LupusTheCanine commented:
What will happen if this is used with ELRS in MSVLink mode?

peterbarker commented:
> What will happen if this is used with ELRS in MSVLink mode?

The ELRS people have been poked to use the correct packet rather than RC overrides.

... and the answer at the moment is "nothing" - so long as they don't try to use this option!  If they do... probably still nothing as the RC inputs are perceived as overrides rather than as stick input.
